### PR TITLE
stm32h7: Fix race condition in ADC interrupt handling

### PR DESCRIPTION
**Pull request by antmerlino**
## Summary
While doing a multiple channel ADC sample, it was possible for a race condition to occur. The fundamental issue is that the driver was explicitly clearing the EOC interrupt bit.  The EOC interrupt bit should only ever be cleared by reading the ADC_DR (Data Register). Here is what was happening:

1) ADC is triggered and the EOC interrupt fires.
2) Inside the ISR, one or more samples are read from the data register.  Each time a read occurs the EOC interrupt is cleared, but may immediately come back up because there is more data to read.
3) At some point, the EOC interrupt falls, but there are still more channels left to sample - that's okay, we will re-enter the interrupt when the EOC bit rises again.
4) We clear ALL interrupts, including the EOC interrupt, indiscriminately.  By doing this we create a race where we've left our loop of sampling channels, the EOC bit rises, and then we clear it. We will now never receive this interrupt.

To fix this we only clear interrupts we service and we NEVER clear the EOC interrupt explicitly. We MUST always clear the EOC interrupt by just reading ADC_DR. 

## Impact
None. Driver fixed.

## Testing
Trigger ADC repeatedly and observe channel shift and corrupt data. Apply fix.  Run for extended period of time with no channel shift and consistently correct data.

### CODE REVIEW

File: arch/arm/src/stm32h7/stm32_adc.c [hunk 1]
```diff
@@ -2008,6 +2010,12 @@ static int adc_interrupt(struct adc_dev_s *dev, uint32_t adcisr)
       /* Stop ADC conversions to avoid continuous interrupts */
 
       adc_startconv(priv, false);
+
+      /* Clear the interrupt. This register only accepts write 1's so its
+       * safe to only set the 1 bit without regard for the rest of the register
```
Comment (acassis) [inline arch/arm/src/stm32h7/stm32_adc.c:2015]: Please fix long line

File: arch/arm/src/stm32h7/stm32_adc.c [hunk 2]
```diff
@@ -2037,6 +2045,10 @@ static int adc_interrupt(struct adc_dev_s *dev, uint32_t adcisr)
           priv->cb->au_reset(dev);
         }
 
+      /* Clear the interrupt. This register only accepts write 1's so its
+       * safe to only set the 1 bit without regard for the rest of the register
```
Comment (acassis) [inline arch/arm/src/stm32h7/stm32_adc.c:2049]: Ditto

File: arch/arm/src/stm32h7/stm32_adc.c [hunk 3]
```diff
@@ -2091,6 +2103,10 @@ static int adc_interrupt(struct adc_dev_s *dev, uint32_t adcisr)
             }
         }
       while ((adc_getreg(priv, STM32_ADC_ISR_OFFSET) & ADC_INT_EOC) != 0);
+
+      /* We dont't add EOC to the bits to clear. It will cause a race condition.
```
Comment (acassis) [inline arch/arm/src/stm32h7/stm32_adc.c:2107]: Please remove dangling space and long line

**Review by jerpelea — APPROVED** [review]
I would appreciate if you add a commit description for future refference

**Pull request merged**